pyscf/mcscf/apc.py

```diff
@@ -0,0 +1,410 @@
+    until the size constraint max_size is met.
+    
+    Args:
+        orbs: 2D Numpy Array
+            Orbitals to choose from, spanning the entire basis (must be square matrix of coefficients)
+        occ: 1D Numpy Array
+            Orbital occupations for orbs (2,1,0); nactel will be set to the number of electrons in the selected orbitals
+        entropies: 1D Numpy Array 
+            Importance measurment used to rank the orbitals
+        max_size: Int or Tuple
+            Active space size constraint. If tuple, interpreted as (nelecas,ncas), if int interpreted as max # of orbitals
+            
+    Returns:
+        active-space-size, #-active-electrons, orbital-initial-guess, chosen-active-orbital-indeces
+        
+    Example:
+    
+    #Randomly ranked orbitals
+    >>> import numpy as np
+    >>> from pyscf import gto, scf, mcscf
+    >>> from pyscf.mcscf import apc
+    >>> mol = gto.M(atom='H 0 0 0; H 0 0 1', basis='ccpvtz')
+    >>> mf = scf.RHF(mol).run()
+    >>> entropies = np.random.choice(np.arange(len(mf.mo_occ)),len(mf.mo_occ),replace=False)
+    >>> chooser = apc.Chooser(mf.mo_coeff,mf.mo_occ,entropies,max_size=(2,2))
+    >>> ncas, nelecas, casorbs, active_idx = chooser.kernel()
+    >>> mc = mcscf.CASSCF(mf, ncas, nelecas).run(casorbs)
+    """
+    
+    def __init__(self,orbs,occ,entropies,max_size=(8,8),verbose=4):
+        #Check that we have a full set of orbitals:
+        assert(orbs.shape[0] == orbs.shape[1])
+        assert(len(occ) == len(entropies))
+        assert(len(entropies) == orbs.shape[1])
+        
+        self.log = logger.new_logger(lib.StreamObject,verbose)
+        self.orbs = orbs
+        self.occ = np.array(occ)
+        self.entropies = entropies
+        self.max_size = max_size
+        self.verbose = verbose
+        
+    def _ncsf(self,nactel,norbs):
+        from scipy.special import comb
+        alpha = int(nactel//2 + nactel%2)
+        beta = int(nactel//2)
+        term1 = comb(norbs,alpha)*comb(norbs,beta)
+        term2 = comb(norbs,alpha+1)*comb(norbs,beta-1)
+        ncsf = term1-term2
+        return ncsf
+        
+    def _calc_ncsf(self,active_idx):
+        occ = self.occ
+        nactel = np.sum(np.array(occ)[active_idx])
+        norbs = len(active_idx)
+        return self._ncsf(nactel,norbs)
+    
+    def _as_is_reasonable(self,active_idx):
+        #Checks active space reasonability
+        occ = self.occ
+        
+        nactel = np.sum(np.array(occ)[active_idx])        
+        num_os = len(np.where(occ == 1)[0])
+        
+        condition1 = (nactel > 0)
+        condition2 = (nactel < 2*len(active_idx))
+        condition3 = (len(active_idx) >= num_os)
+        
+        if (condition1 and condition2 and condition3):
+            return True
+        else:
+            self.log.debug("Active space is not reasonable!")
+            if not condition1:
+                self.log.debug("Condition 1 not met")
```

Comment:
logger print also `nactel, len(active_idx), num_os` as the reason why conditions are not met.

Reply:
Done

Reply:
Added 👍